From 0521fb9e9a0b4ed024cd09f26cc0fe22b47aaf51 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sun, 29 Oct 2017 18:41:48 -0400
Subject: [PATCH] Add option to make parrots stay on shoulders despite movement

Makes parrots not fall off whenever the player changes height, or touches water, or gets hit by a passing leaf.
Instead, switches the behavior so that players have to sneak to make the birds leave.

To be removed in the future (1.13?)
This can be accomplished via plugins

diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 33f097c..c05afe0 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -410,4 +410,10 @@ public class PaperWorldConfig {
         maxCollisionsPerEntity = getInt("max-entity-collisions", this.spigotConfig.getInt("max-entity-collisions", 8));
         log("Max Entity Collisions: " + maxCollisionsPerEntity);
     }
+
+    public boolean parrotsHangOnBetter;
+    private void parrotsHangOnBetter() {
+        parrotsHangOnBetter = getBoolean("parrots-are-unaffected-by-player-movement", false);
+        log("Parrots are unaffected by player movement: " + parrotsHangOnBetter);
+    }
 }
diff --git a/src/main/java/net/minecraft/entity/player/EntityPlayer.java b/src/main/java/net/minecraft/entity/player/EntityPlayer.java
index d50be00..bbf5be1 100644
--- a/src/main/java/net/minecraft/entity/player/EntityPlayer.java
+++ b/src/main/java/net/minecraft/entity/player/EntityPlayer.java
@@ -490,7 +490,7 @@ public abstract class EntityPlayer extends EntityLivingBase {
         this.playShoulderEntityAmbientSound(this.getRightShoulderEntity());
 
         if (!this.world.isRemote && (this.fallDistance > 0.5F || this.isInWater() || this.isRiding()) || this.capabilities.isFlying) {
-            this.spawnShoulderEntities();
+            if (!this.world.paperConfig.parrotsHangOnBetter) this.spawnShoulderEntities();  // Paper - Hang on!
         }
     }
 
diff --git a/src/main/java/net/minecraft/network/NetHandlerPlayServer.java b/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
index fb17aa9..08571e4 100644
--- a/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
+++ b/src/main/java/net/minecraft/network/NetHandlerPlayServer.java
@@ -1661,6 +1661,11 @@ public class NetHandlerPlayServer implements INetHandlerPlayServer, ITickable {
             this.player.markPlayerActive();
             switch (packetIn.getAction()) {
                 case START_SNEAKING:
+                    // Paper start - Get off!
+                    if (this.player.world.paperConfig.parrotsHangOnBetter) {
+                        this.player.spawnShoulderEntities();
+                    }
+                    // Paper end
                     this.player.setSneaking(true);
                     break;
                 case STOP_SNEAKING:
-- 
2.18.0

